kstatus/README.md

```diff
@@ -0,0 +1,116 @@
+insufficient progress (timeout).
+
+The use of the "abnormal-true" pattern has some challenges. If the controller is not running, or for some
+reason not able to update the resource, it will look like it is in a good state when that is not true. The
+solution to this issue is to adopt the pattern used by several of the built-in types where there is an
+`observedGeneration` property on the status object which is set by the controller during the reconcile loop.
+If the `generation` and the `observedGeneration` of a resource does not match, it means there are changes
+that the controller has not yet seen, and therefore not acted upon.
+
+## Features
+
+The library is currently separated into two packages, one that provides the basic functionality, and another that
+builds upon the basics to provide a higher level API.
+
+**sigs.k8s.io/kustomize/kstatus/status**: Provides two basic functions. First, it provides the `Compute` function
+that takes a single resource and computes the status for this resource based on the fields in the status object for
+the resource. Second, it provides the `Augment` function that computes the appropriate standard conditions based on
+the status object and then amends them to the conditions in the resource. Both of these functions currently operate
+on Unstructured types, but this should eventually be changed to rely on the kyaml library. Both of these functions 
+compute the status and conditions solely based on the data in the resource passed in. It does not communicate with
+a cluster to get the latest state of the resources.
+
+**sigs.k8s.io/kustomize/kstatus/wait**: This package builds upon the status package and provides functionality that
+will fetch the latest state from a cluster. It provides the `FetchAndResolve` function that takes list of resource
+identifiers, fetches the latest state for all the resources from the cluster, computes the status for all of them and
+returns the results. `WaitForStatus` accepts a list of resource identifiers and will poll cluster for the status of 
+the resources until all resources have reached the `Current` status.
+
+## Challenges
+
+### Status is not obvious for all resource types
+
+For some types of resources, it is pretty clear what the different statuses mean. For others, it
+is far less obvious. For example, what does it mean that a PodDisruptionBudget is Current? Based on
+the assumptions above it probably should be whenever the controller has observed the resource
+and updated the status object of the PDB with information on allowed disruptions. But currently, a PDB is
+considered Current when the number of healthy replicas meets the threshold given in the PDB. Also, should
+the presence of a PDB influence when a Deployment is considered Current? This would mean that a Deployment
+should be considered Current whenever the number of replicas reach the threshold set by the corresponding
+PDB. This is not currently supported as described below.
```

Comment:
Isn't PDB like a ConfigMap in the sense that a Deployment looks at it for instructions?  Its the Deployment that changes, not the PDB.  We only care about the status of the Deployment.  Seems find to have a a class of objects that simply don't have interesting status/conditions, but still 'implement' some interface so we can just for their status anyway.

Reply:
That's a nice discussion.  kinda want to add it in somehow to further color the difficultly/scope of the problem being solved here. up you